mmv1/third_party/terraform/utils/framework_location.go

```diff
@@ -0,0 +1,99 @@
+package google
+
+import (
+	"errors"
+	"fmt"
+
+	"github.com/hashicorp/terraform-plugin-framework/types"
+)
+
+type LocationDescriber interface {
+	getLocationDescription(providerConfig *frameworkProvider) LocationDescription
+}
+
+type LocationDescription struct {
+	LocationSchemaField types.String
+	RegionSchemaField   types.String
+	ZoneSchemaField     types.String
+
+	ResourceLocation types.String
+	ResourceRegion   types.String
+	ResourceZone     types.String
+
+	ProviderRegion types.String
+	ProviderZone   types.String
+}
+
+func (ld *LocationDescription) getLocation() (types.String, error) {
+	// Location from resource config
+	if !ld.ResourceLocation.IsNull() && !ld.ResourceLocation.IsUnknown() {
+		return ld.ResourceLocation, nil
+	}
+
+	// Location from region in resource config
+	if !ld.ResourceRegion.IsNull() && !ld.ResourceRegion.IsUnknown() {
+		return ld.ResourceRegion, nil
+	}
+
+	// Location from zone in resource config
+	if !ld.ResourceZone.IsNull() && !ld.ResourceZone.IsUnknown() {
+		location := GetResourceNameFromSelfLink(ld.ResourceZone.ValueString()) // Zone could be a self link
+		return types.StringValue(location), nil
+	}
+
+	// Location from zone in provider config
+	if !ld.ProviderZone.IsNull() && !ld.ProviderZone.IsUnknown() {
+		return ld.ProviderZone, nil
+	}
+
+	return types.StringNull(), nil
+}
+
+func (ld *LocationDescription) getRegion() (types.String, error) {
```

Comment:
it looks like this doesn't replicate the GKE-specific behavior; I am guessing that the idea is that we would rewrite GKE in such a way that the workaround isn't necessary? (I like that idea, just want to confirm!)

Reply:
tldr; I think this is a problem to tackle closer to the time that these edge case resources are migrated.

---

From a quick search of the codebase where `getRegionFromSchema` is used (and `parseRegionalFieldValue`, which passes arguments into that function) I think that edge case is relevant in only a few places. I've almost certainly missed a few, but it's not very widespread from a first glace:

[google_redis_instance](https://github.com/hashicorp/terraform-provider-google/blob/14c4df39c94c4c1bc675241fbaf4376e4a5c8a8e/google/resource_redis_instance.go#L1966)

```go
region, err := getRegionFromSchema("region", "location_id", d, config)
```

[google_container_cluster](https://github.com/hashicorp/terraform-provider-google-beta/blob/4c61eb4c660d2379907224a3938b078b849a845c/google-beta/resource_container_cluster.go#L2056)

```go
subnetwork, err := parseRegionalFieldValue("subnetworks", v.(string), "project", "location", "location", d, config, true) // variant of ParseSubnetworkFieldValue
```

For `google_redis_instance` I think custom code is the answer. Either something new, or to implement the `getLocationDescription` function with `location_id` being used as the zone value (as currently a value of region is used over location_id).

For things that use `parseRegionalFieldValue` I think we need to tackle that when we move on to migrating that function. In the case above for `google_container_cluster` I think that it would need to have custom code that implements the `getLocationDescription` function and uses the value of the location field to set location, region and zone within the `LocationDescription` struct?

Reply:
Please let me know if I'm overlooking something that I lack context for though! It's hard to foresee and unpick edge cases, especially if the current solution for those edge cases relies on features of the SDK that are missing in the plugin framework (accessing resource data via getters that can take string variables to identify fields with).

Reply:
Looking over the calling code - I think this code would only get triggered for google_container_cluster. Redis does have non-standard behavior but it still has different names for the two fields so wouldn't trigger this code block. I agree that the design you're proposing would let us handle this case in `GetLocationDescription` for the GKE resources. I had been worried that there might be non-obvious edge cases but it looks like everything is using hardcoded strings - so this seems like a fine way to proceed 👍  thanks for looking into it!